src/arch/mips/linux/linux.hh

```diff
@@ -192,6 +203,8 @@ class MipsLinux : public Linux, public OpenFlagTable<MipsLinux>
        uint32_t mem_unit;  /* Memory unit size in bytes */
     };
 
+    typedef uint32_t size_t;
+    typedef int32_t off_t;
```

Comment:
Why this change? Has it been tested? If so, how was it tested?

Reply:
This set of changes is needed because the `lseekFunc` and `getcwdFunc` have been templatised with the OS type parameter, and take an off_t or size_t type parameter. This is common practice for all syscalls that require parameters of this type (`mmapFunc` is a good example). As a result, all the ABIs built with syscall tables that include these functions must pass as a template parameter an OS that defines size_t and off_t types (this is, as I said, alreadt common practice - see gem5::X86Linux64 for an example). `lseek` and `getcwd` were quite ubiquitous, and it would be unfortunate to remove them from the table of supported syscalls for any of the other architectures, hence the need for the typedefs.

Reply:
I'm still concerned that it isn't tested. I agree that we don't want to remove support for `lseek` and `getcwd`, but I also don't want to have any regressions in support.

Is there any way you can test this?

Reply:
I think @powerjg concerns for testing are valid. However IMHO the MIPS/POWER/SPARC change  is relatively straightforward; it is simply replacing an hardcoded(uint64_t) or host dependent (unsigned long) argument with the appropriate ISA size. 

That sizes are properly chosen can be verified in godbolt, choosing different sparc/mips/power cross compiler
```

#include <sys/types.h>
#include <iostream>
#include <stdint.h>

struct CurrISA
{
    using expected_off_t = uint32_t;
    using expected_size_t = uint32_t;
};

template <class ISA>
void checker()
{
    static_assert(
        sizeof(off_t) == sizeof(typename ISA::expected_off_t));
    static_assert(
        sizeof(size_t) == sizeof(typename ISA::expected_size_t));
}

int main()
{
    checker<CurrISA>();
}
```

Let me know if this is enough for you. We can try to come up with lseek testing programs but as you can imagine cross-compiling these not so used ISAs (I am sure our gem5 testing of MIPS SE is close to zero) might not be the simplest thing to do :)

Reply:
Yeah, this is great. I agree that it's a pain to test. I wish I could convince people to get rid of some of these rarer ISAs...

The other thing that I would like to make sure is that all of the code compiles. I *think* if you compile `ALL` it should compile all of this code, but it would be good to double check.

Reply:
The code compiles when building the ALL target. I have tried using this test program on PowerPC, using the cross compiler provided in the Docker image in tests/test-progs/hello/src and the cross compiler available for Ubuntu 22.01 in the apt package manager (a.k.a. powerpc64le-linux-gnu-gcc):
`#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/syscall.h>
#include <errno.h>

int main() {  
    int fd = open("example.txt", O_RDONLY | O_CREAT, 0644);  
    if (fd == -1) {  
        perror("open");  
        return 1;  
    }
    off_t offset = syscall(SYS_lseek, fd, 10, SEEK_SET);
    if (offset == -1) {
        perror("lseek");  
        close(fd);  
        return 1;  
    }  
    char buffer[10];  
    ssize_t bytesRead = read(fd, buffer, sizeof(buffer));  
    if (bytesRead == -1) {  
        perror("read");  
        close(fd);  
        return 1;  
    }  
    printf("Read %zd bytes: %.*s\n", bytesRead, (int)bytesRead, buffer);  
    close(fd);  
    return 0;
}
`
When run under gem5, both versions of the produced binary encounter unimplemented instructions which are unrelated to syscall emulation and their implementation is outside the scope of this bugfix.
If you don't feel comfortable committing the patch in this form, an alternative would be to keep around an old implementation of the affected syscalls and revert the type definition changes for these rarer ISAs. This would ensure the exact same state prior to the patches is preserved.

Reply:
I guess this is OK as the patch currently is. If something ends up breaking, we will at least have this conversation to come back to. I just feel uncomfortable committing untested code.